### PR TITLE
GPU extruded operator tests

### LINKED ISSUE #1169: Test GPU spectral element operators on extruded box and spherical spaces
Test that existing GPU spectral element operators work on extruded spherical spaces.

- It should test both extruded box and sphere
- It should test the common operators (div, grad, curl), and some compositions (e.g. `div.(grad.(x))`)
  - if you hit #1226, then mark the test as broken.

### PR DESCRIPTION
**Pull request by valeriabarra**
This WIP PR will close #1169 (when ready and all issues in the GPU kernels are solved). 

Content:
 
- [x] Added a test to test `div` operator (Works!)

PR Review:
- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: test/Operators/hybrid/extruded_cuda.jl [hunk 1]
```diff
@@ -0,0 +1,77 @@
+#=
```
Comment (charleskawczynski) [inline test/Operators/hybrid/extruded_cuda.jl:1]: Can we remove some boiler plate and define 

```julia
function get_space(context)
    FT = Float64

    # Define vert domain and mesh
    vertdomain = Domains.IntervalDomain(
        Geometry.ZPoint{FT}(0.0),
        Geometry.ZPoint{FT}(1.0);
        boundary_tags = (:bottom, :top),
    )
    vertmesh = Meshes.IntervalMesh(vertdomain, nelems = 10)

    # Define vert topology and space
    verttopology = Topologies.IntervalTopology(cpu_context, vertmesh)
    vert_center_space = Spaces.CenterFiniteDifferenceSpace(verttopology)

    # Define horz domain and mesh
    horzdomain = Domains.SphereDomain(FT(30.0))
    horzmesh = Meshes.EquiangularCubedSphere(horzdomain, 4)

    quad = Spaces.Quadratures.GLL{3 + 1}()

    # Define horz topology and space
    horztopology = Topologies.Topology2D(cpu_context, horzmesh)
    horzspace = Spaces.SpectralElementSpace2D(horztopology, quad)

    # Define hv space
    hv_center_space = Spaces.ExtrudedFiniteDifferenceSpace(
        horzspace,
        vert_center_space,
    )
    return hv_center_space
end
```
?

File: test/Operators/hybrid/extruded_cuda.jl [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    gpu_context = SingletonCommsContext(ClimaComms.CUDADevice())
+    device = ClimaComms.device() #ClimaComms.CUDADevice()
+    println("running test on $device device")
+
+
+    # Define hv CPU space
+    hv_center_space_cpu = get_space(cpu_context)
+
+    # Define hv GPU space
+    hv_center_space_cpu = get_space(cpu_context)
```
Comment (charleskawczynski) [inline test/Operators/hybrid/extruded_cuda.jl:55]:
```suggestion
    hv_center_space_gpu = get_space(gpu_context)
```

**Comment by simonbyrne** [comment]
The file should be `include`-ed in the `runtests` (in the GPU block)

File: test/Operators/hybrid/extruded_cuda.jl [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    coords_cpu = Fields.coordinate_field(hv_center_space_cpu)
+    coords_gpu = Fields.coordinate_field(hv_center_space_gpu)
+
+    # Define CPU & GPU fields
+    x_cpu = Geometry.UVWVector.(cosd.(coords_cpu.lat), 0.0, 0.0)
+    x_gpu = Geometry.UVWVector.(cosd.(coords_gpu.lat), 0.0, 0.0)
+
+    # Test div operator
+    div = Operators.Divergence()
+    @test parent(div.(x_cpu)) ≈ Array(parent(div.(x_gpu)))
```
Comment (simonbyrne) [inline test/Operators/hybrid/extruded_cuda.jl:67]: Can we test other operators (`grad`, `curl`), and some composed operators (`div.(grad.(x))`)?

Comment (valeriabarra) [reply test/Operators/hybrid/extruded_cuda.jl:67]: Yes, absolutely. I had left a [comment](https://github.com/CliMA/ClimaCore.jl/issues/1169#issuecomment-1535552647) in the related Issue

File: test/Operators/hybrid/extruded_sphere_cuda.jl [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    @test parent(wcurl.(Geometry.Covariant12Vector.(g_cpu))) ≈
+          Array(parent(wcurl.(Geometry.Covariant12Vector.(g_gpu))))
+    @test parent(wcurl.(Geometry.Covariant3Vector.(f_cpu))) ≈
+          Array(parent(wcurl.(Geometry.Covariant3Vector.(f_gpu))))
+
+    # Test wdiv(grad()) composed Laplace operator
+    grad = Operators.Gradient()
+    @test parent(wdiv.(grad.(f_cpu))) ≈ Array(parent(wdiv.(grad.(f_gpu))))
+    @test_broken parent(wdiv.(grad.(x_cpu))) ≈
+                 Array(parent(wdiv.(grad.(x_gpu))))
```
Comment (valeriabarra) [inline test/Operators/hybrid/extruded_sphere_cuda.jl:144]: I added the requested spherical tests (and now will move on to work on the 3d hybrid box ones). 

@simonbyrne 
I found that the `grad.(x_gpu)` does not work, with `x_gpu` defined as above (`UVWVector`), but it does work with `f_gpu`, defined as a scalar field. Is this expected? The corresponding CPU version works with both.

Comment (valeriabarra) [reply test/Operators/hybrid/extruded_sphere_cuda.jl:144]: Same thing happens in the extruded 3d box case (now completed)

**Comment by valeriabarra** [comment]
@simonbyrne , @charleskawczynski , I believe I added all requested tests.
This PR is ready for review. Thanks!

File: test/Operators/hybrid/extruded_3dbox_cuda.jl [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+            cos.(coords_gpu.z),
+        )
+
+    # Test grad operator
+    grad = Operators.Gradient()
+    @test parent(grad.(f_cpu)) ≈ Array(parent(grad.(f_gpu)))
+
+    # Test div operator
+    div = Operators.Divergence()
+    @test parent(div.(x_cpu)) ≈ Array(parent(div.(x_cpu)))
```
Comment (charleskawczynski) [inline test/Operators/hybrid/extruded_3dbox_cuda.jl:97]:
```suggestion
    @test parent(div.(x_cpu)) ≈ Array(parent(div.(x_gpu)))
```

File: test/Operators/hybrid/extruded_sphere_cuda.jl (hunk 4 above)
Comment (simonbyrne) [reply test/Operators/hybrid/extruded_sphere_cuda.jl:144]: yes, you can only take a grad of a scalar quantity for now.

**Review by simonbyrne — APPROVED** [review]

**Pull request merged**